File: Java/61-RotateList.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if (head == null || k == 0 || head.next == null)
            return head;
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode slow = dummy, fast = dummy;
        int listLen = 0;
        while (fast.next != null){
            fast = fast.next;
            listLen++;
        }
        for (int i = 0; i < listLen - (k % listLen); i++){
            slow = slow.next;
        }
        fast.next = dummy.next;
        dummy.next = slow.next;
        slow.next = null;
        return dummy.next;
    }
}

File: cpp/61-RotateList.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
 //Given 1->2->3->4->5->NULL and k = 2,
 //return 4->5->1->2->3->NULL.

class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if (!head || !head->next || k == 0)
            return head;
        ListNode* curr = head, *newHead = NULL;
        int len = 1;
        while (curr->next) {
            curr = curr->next;
            len++;
        }
        //cout << len;
        k = len - k % len;
        curr->next = head;
        //wrong:
        //curr = curr->next;

        while (curr && k > 0) {
            curr = curr->next;
            k--;
        }
        head = curr->next;
        //VERY IMPORTANT!!
        curr->next = NULL;
        return head;
    }
};
